File: rebrand.py
#!/usr/bin/env python3
"""
Universal Rebrand Script - Agent Skill Kit
===========================================
Automatically rebrand entire codebase with intelligent pattern matching.

Usage:
    python rebrand.py "Old Name" "New Name" [--dry-run]
    
Example:
    python rebrand.py "Antigravity Kit" "Agent Skill Kit"
    python rebrand.py "AgentSkillKit" "SuperAI" --dry-run
"""

import re
import sys
import argparse
from pathlib import Path
from typing import List, Dict, Tuple
import json

class SmartRebrander:
    def __init__(self, old_name: str, new_name: str, root_dir: Path):
        self.old_name = old_name
        self.new_name = new_name
        self.root_dir = root_dir
        
        # Generate all name variations
        self.variations = self._generate_variations()
        
        # Files to scan
        self.file_patterns = ["*.py", "*.md", "*.json", "*.js", "*.ts", "*.tsx", "*.jsx"]
        
        # Exclusions
        self.exclude_dirs = {".git", "node_modules", ".next", "dist", "build", "__pycache__", ".venv"}
        self.exclude_files = {"package-lock.json", "yarn.lock", "pnpm-lock.yaml"}
        
        # Results
        self.changes = []
        self.errors = []
        
    def _generate_variations(self) -> Dict[str, str]:
        """Generate all naming convention variations"""
        # Helper functions
        def to_kebab(text): return text.lower().replace(" ", "-")
        def to_snake(text): return text.lower().replace(" ", "_").replace("-", "_")
        def to_camel(text): return ''.join(word.capitalize() for word in text.split())
        def to_pascal(text): return to_camel(text)
        def to_lower(text): return text.lower().replace(" ", "")
        def to_upper(text): return text.upper().replace(" ", "_")
        
        variations = {
            # Original
            "original": (self.old_name, self.new_name),
            
            # Case variations
            "lowercase": (self.old_name.lower(), self.new_name.lower()),
            "uppercase": (self.old_name.upper(), self.new_name.upper()),
            
            # Naming conventions
            "kebab-case": (to_kebab(self.old_name), to_kebab(self.new_name)),
            "snake_case": (to_snake(self.old_name), to_snake(self.new_name)),
            "camelCase": (to_camel(self.old_name), to_camel(self.new_name)),
            "PascalCase": (to_pascal(self.old_name), to_pascal(self.new_name)),
            
            # Special formats
            "no-spaces-lower": (to_lower(self.old_name), to_lower(self.new_name)),
            "SCREAMING_SNAKE": (to_upper(self.old_name), to_upper(self.new_name)),
        }
        
        return variations
    
    def find_files(self) -> List[Path]:
        """Find all files to process"""
        files = []
        
        for pattern in self.file_patterns:
            for file in self.root_dir.rglob(pattern):
                # Skip excluded directories
                if any(excluded in file.parts for excluded in self.exclude_dirs):
                    continue
                
                # Skip excluded files
                if file.name in self.exclude_files:
                    continue
                
                files.append(file)
        
        return files
    
    def process_file(self, file_path: Path, dry_run: bool = False) -> Dict:
        """Process a single file"""
        try:
            # Read file
            content = file_path.read_text(encoding='utf-8')
            original_content = content
            
            # Track changes for this file
            file_changes = []
            
            # Apply all variations
            for variant_name, (old_variant, new_variant) in self.variations.items():
                # Skip if old variant not in content
                if old_variant not in content:
                    continue
                
                # Count occurrences
                count = content.count(old_variant)
                
                # Context-aware replacement
                # Don't replace inside URLs, git commits, etc.
                if self._should_replace(content, old_variant, file_path):
                    content = content.replace(old_variant, new_variant)
                    file_changes.append({
                        "variant": variant_name,
                        "old": old_variant,
                        "new": new_variant,
                        "count": count
                    })
            
            # Write back if changed
            if content != original_content:
                if not dry_run:
                    file_path.write_text(content, encoding='utf-8')
                
                return {
                    "file": str(file_path.relative_to(self.root_dir)),
                    "changes": file_changes,
                    "modified": not dry_run
                }
            
            return None
            
        except Exception as e:
            self.errors.append({
                "file": str(file_path.relative_to(self.root_dir)),
                "error": str(e)
            })
            return None
    
    def _should_replace(self, content: str, old_text: str, file_path: Path) -> bool:
        """Determine if we should replace in this context"""
        # Skip if in URLs (heuristic)
        url_patterns = [
            r'https?://[^\s]*' + re.escape(old_text),
            r'git\+https://[^\s]*' + re.escape(old_text),
        ]
        
        for pattern in url_patterns:
            if re.search(pattern, content, re.IGNORECASE):
                # If it's a URL, only replace if explicitly domain-related
                if file_path.suffix not in ['.md', '.txt']:
                    return False
        
        # Skip if in git commit hashes (SHA patterns)
        if re.match(r'^[a-f0-9]{7,40}$', old_text):
            return False
        
        # Otherwise safe to replace
        return True
    
    def run(self, dry_run: bool = False):
        """Execute rebrand"""
        print("🚀 Starting Smart Rebrand")
        print(f"   Old Name: {self.old_name}")
        print(f"   New Name: {self.new_name}")
        print(f"   Mode: {'DRY RUN' if dry_run else 'LIVE'}")
        print()
        
        # Show variations
        print("📝 Name Variations:")
        for variant_name, (old, new) in self.variations.items():
            if old != new:  # Only show if different
                print(f"   {variant_name:20s}: {old:30s} → {new}")
        print()
        
        # Find files
        files = self.find_files()
        print(f"🔍 Scanning {len(files)} files...\n")
        
        # Process each file
        for file in files:
            result = self.process_file(file, dry_run)
            if result:
                self.changes.append(result)
        
        # Print results
        self._print_results(dry_run)
    
    def _print_results(self, dry_run: bool):
        """Print summary results"""
        print("\n" + "="*70)
        print("📊 REBRAND SUMMARY")
        print("="*70)
        
        if not self.changes:
            print("\n✅ No changes needed - all references already use new name!")
            return
        
        # Group by file
        print(f"\n📝 Modified Files: {len(self.changes)}")
        print("-" * 70)
        
        total_replacements = 0
        for change in self.changes:
            file_name = change['file']
            changes = change['changes']
            
            file_total = sum(c['count'] for c in changes)
            total_replacements += file_total
            
            print(f"\n{file_name} ({file_total} replacements)")
            for c in changes:
                print(f"  • {c['variant']:15s}: {c['old']:25s} → {c['new']} ({c['count']}x)")
        
        # Errors
        if self.errors:
            print("\n" + "="*70)
            print("❌ ERRORS")
            print("="*70)
            for error in self.errors:
                print(f"  {error['file']}: {error['error']}")
        
        # Summary
        print("\n" + "="*70)
        print(f"Total Replacements: {total_replacements}")
        print(f"Files Modified: {len(self.changes)}")
        print(f"Errors: {len(self.errors)}")
        print("="*70)
        
        if dry_run:
            print("\n⚠️  DRY RUN - No files were modified")
            print("   Run without --dry-run to apply changes")
        else:
            print("\n✅ Rebrand complete!")
            print("   Don't forget to:")
            print("   1. Review changes with: git diff")
            print("   2. Test the application")
            print("   3. Commit: git add . && git commit -m 'rebrand: ...'")
    
    def export_report(self, filename: str = "rebrand_report.json"):
        """Export detailed report"""
        report = {
            "old_name": self.old_name,
            "new_name": self.new_name,
            "variations": {k: {"old": v[0], "new": v[1]} for k, v in self.variations.items()},
            "summary": {
                "files_modified": len(self.changes),
                "total_replacements": sum(
                    sum(c['count'] for c in change['changes'])
                    for change in self.changes
                ),
                "errors": len(self.errors)
            },
            "changes": self.changes,
            "errors": self.errors
        }
        
        with open(filename, 'w') as f:
            json.dump(report, f, indent=2)
        
        print(f"\n📄 Detailed report saved to: {filename}")


def main():
    parser = argparse.ArgumentParser(
        description="Smart rebrand tool for entire codebase",
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog="""
Examples:
  python rebrand.py "Antigravity Kit" "Agent Skill Kit" --dry-run
  python rebrand.py "OldCompany" "NewCompany"
  python rebrand.py "Legacy AI" "Modern AI" --export-report
        """
    )
    
    parser.add_argument("old_name", help="Old name to replace")
    parser.add_argument("new_name", help="New name to use")
    parser.add_argument("--dry-run", action="store_true", help="Preview changes without modifying files")
    parser.add_argument("--export-report", action="store_true", help="Export detailed JSON report")
    parser.add_argument("--root", default=".", help="Root directory to search (default: current)")
    
    args = parser.parse_args()
    
    # Validate
    if args.old_name == args.new_name:
        print("❌ Error: Old name and new name are the same!")
        sys.exit(1)
    
    # Confirm if not dry run
    if not args.dry_run:
        print(f"⚠️  You are about to rebrand '{args.old_name}' to '{args.new_name}'")
        print("   This will modify files in:", Path(args.root).resolve())
        response = input("\n   Continue? (yes/no): ")
        if response.lower() not in ['yes', 'y']:
            print("Aborted.")
            sys.exit(0)
        print()
    
    # Execute
    root = Path(args.root).resolve()
    rebrander = SmartRebrander(args.old_name, args.new_name, root)
    rebrander.run(dry_run=args.dry_run)
    
    # Export report if requested
    if args.export_report:
        rebrander.export_report()


if __name__ == "__main__":
    main()

File: rebrand.mjs
#!/usr/bin/env node
/**
 * Smart Rebrand CLI - Agent Skill Kit
 * Beautiful CLI rebrand tool powered by @clack/prompts
 */

import * as p from '@clack/prompts';
import { readdir, readFile, writeFile } from 'fs/promises';
import { join, relative, extname } from 'path';
import { glob } from 'glob';
import color from 'picocolors';

class RebrandCLI {
    constructor() {
        this.variations = new Map();
        this.changes = [];
        this.errors = [];
    }

    generateVariations(oldName, newName) {
        const toKebab = (text) => text.toLowerCase().replace(/[\s_]/g, '-');
        const toSnake = (text) => text.toLowerCase().replace(/[\s-]/g, '_');
        const toCamel = (text) => {
            const words = text.split(/[\s-_]+/);
            return words[0].toLowerCase() + words.slice(1).map(w => w.charAt(0).toUpperCase() + w.slice(1).toLowerCase()).join('');
        };
        const toPascal = (text) => {
            const words = text.split(/[\s-_]+/);
            return words.map(w => w.charAt(0).toUpperCase() + w.slice(1).toLowerCase()).join('');
        };
        const toConstant = (text) => text.toUpperCase().replace(/[\s-]/g, '_');

        return new Map([
            ['original', [oldName, newName]],
            ['lowercase', [oldName.toLowerCase(), newName.toLowerCase()]],
            ['uppercase', [oldName.toUpperCase(), newName.toUpperCase()]],
            ['kebab-case', [toKebab(oldName), toKebab(newName)]],
            ['snake_case', [toSnake(oldName), toSnake(newName)]],
            ['camelCase', [toCamel(oldName), toCamel(newName)]],
            ['PascalCase', [toPascal(oldName), toPascal(newName)]],
            ['CONSTANT_CASE', [toConstant(oldName), toConstant(newName)]],
        ]);
    }

    async findFiles() {
        const patterns = ['**/*.{js,ts,tsx,jsx,py,md,json}'];
        const ignore = ['**/node_modules/**', '**/.git/**', '**/dist/**', '**/build/**'];

        return await glob(patterns, { ignore, nodir: true });
    }

    async scanFile(filePath) {
        try {
            const content = await readFile(filePath, 'utf-8');
            const refs = [];

            for (const [variant, [oldVal, newVal]] of this.variations) {
                if (content.includes(oldVal)) {
                    const count = (content.match(new RegExp(oldVal.replace(/[.*+?^${}()|[\]\\]/g, '\\$&'), 'g')) || []).length;
                    refs.push({ variant, old: oldVal, new: newVal, count });
                }
            }

            return refs.length > 0 ? { file: filePath, refs } : null;
        } catch (error) {
            this.errors.push({ file: filePath, error: error.message });
            return null;
        }
    }

    async processFile(filePath, dryRun = false) {
        try {
            let content = await readFile(filePath, 'utf-8');
            const original = content;

            for (const [_, [oldVal, newVal]] of this.variations) {
                content = content.replaceAll(oldVal, newVal);
            }

            if (content !== original) {
                if (!dryRun) {
                    await writeFile(filePath, content, 'utf-8');
                }
                return true;
            }
            return false;
        } catch (error) {
            this.errors.push({ file: filePath, error: error.message });
            return false;
        }
    }

    async updatePackageJson(oldName, newName) {
        try {
            const pkgPath = join(process.cwd(), 'package.json');
            const content = await readFile(pkgPath, 'utf-8');
            const pkg = JSON.parse(content);

            // Update description if it contains the old brand name
            if (pkg.description && pkg.description.includes(oldName)) {
                pkg.description = pkg.description.replace(oldName, newName);

                // Write back with pretty formatting
                await writeFile(pkgPath, JSON.stringify(pkg, null, 2) + '\n', 'utf-8');
                return true;
            }

            return false;
        } catch (error) {
            this.errors.push({ file: 'package.json', error: error.message });
            return false;
        }
    }

    async detectCurrentBrand() {
        try {
            const pkgPath = join(process.cwd(), 'package.json');
            const content = await readFile(pkgPath, 'utf-8');
            const pkg = JSON.parse(content);

            // Priority 1: description (more accurate)
            if (pkg.description) {
                const match = pkg.description.match(/^(?:The )?(.+?)(?:\s+for|$)/);
                if (match) return match[1];
            }

            // Priority 2: package name
            if (pkg.name) {
                const match = pkg.name.match(/[@\/]?([^\/]+)(?:\/([^\/]+))?$/);
                if (match) {
                    const name = match[2] || match[1];
                    return name.split('-').map(w => w.charAt(0).toUpperCase() + w.slice(1)).join(' ');
                }
            }

            return null;
        } catch {
            return null;
        }
    }

    async run() {
        console.clear();

        p.intro(color.bgCyan(color.black(' Smart Rebrand v2 ')));

        // Auto-detect current brand
        const currentBrand = await this.detectCurrentBrand();

        if (currentBrand) {
            p.note(
                `Current: ${color.cyan(currentBrand)}\n` +
                color.dim('Detected from package.json'),
                'Current Brand'
            );
        }

        const project = await p.group(
            {
                newName: () => p.text({
                    message: currentBrand
                        ? `What would you like to change "${color.cyan(currentBrand)}" to?`
                        : 'What is the new brand name?',
                    placeholder: 'Agent Skill Kit',
                    validate: (value) => {
                        if (!value) return 'Please enter the new brand name';
                        if (currentBrand && value === currentBrand) return 'New name must be different from current';
                    },
                }),
                mode: () => p.select({
                    message: 'Run mode:',
                    options: [
                        { value: 'dry-run', label: 'Dry Run (preview only)', hint: 'Safe preview' },
                        { value: 'live', label: 'Live Mode (apply changes)', hint: 'Make changes' },
                    ],
                }),
            },
            {
                onCancel: () => {
                    p.cancel('Operation cancelled.');
                    process.exit(0);
                },
            }
        );

        const oldName = currentBrand || await p.text({
            message: 'Could not detect current brand. Please enter manually:',
            validate: (value) => {
                if (!value) return 'Please enter the current brand name';
            },
        });

        if (oldName === project.newName) {
            p.cancel('Old and new names are identical!');
            process.exit(1);
        }

        // Confirm if live mode
        if (project.mode === 'live') {
            const confirmed = await p.confirm({
                message: `Replace "${oldName}" with "${project.newName}" in all files?`,
                initialValue: false,
            });

            if (!confirmed) {
                p.cancel('Aborted.');
                process.exit(0);
            }
        }

        // Generate variations
        this.variations = this.generateVariations(oldName, project.newName);

        // Scan phase
        const s = p.spinner();
        s.start('Scanning files...');

        const files = await this.findFiles();
        const references = [];

        for (const file of files) {
            const refs = await this.scanFile(file);
            if (refs) {
                references.push(refs);
            }
        }

        s.stop(`Found ${references.length} files with references`);

        if (references.length === 0) {
            p.note('No references found. Codebase is already clean!', 'All Clear ✓');
            p.outro(color.green('Nothing to do!'));
            return;
        }

        // Show preview
        p.note(
            references.slice(0, 5).map(r =>
                `${color.dim(r.file)} ${color.yellow(`(${r.refs.length} matches)`)}`
            ).join('\n') +
            (references.length > 5 ? `\n${color.dim(`... and ${references.length - 5} more`)}` : ''),
            'Files to modify'
        );

        if (project.mode === 'dry-run') {
            p.outro(color.yellow('Dry run complete. No files were modified.'));
            return;
        }

        // Apply changes
        s.start('Applying changes...');

        for (const { file } of references) {
            const success = await this.processFile(file, false);
            if (success) {
                this.changes.push(file);
            }
        }

        s.stop('Changes applied');

        // Update package.json with new brand name
        const pkgUpdated = await this.updatePackageJson(oldName, project.newName);
        if (pkgUpdated) {
            p.note(
                `Updated description: "${project.newName}"`,
                color.green('✓ package.json updated')
            );
        }

        // Report
        if (this.errors.length > 0) {
            p.note(
                this.errors.slice(0, 3).map(e =>
                    `${color.red('✗')} ${e.file}\n  ${color.dim(e.error)}`
                ).join('\n'),
                color.red('Errors')
            );
        }

        const nextSteps = `
${color.cyan('1.')} Review changes: ${color.bold('git diff')}
${color.cyan('2.')} Test your application
${color.cyan('3.')} Commit: ${color.bold('git add . && git commit -m "rebrand: ..."')}
        `.trim();

        p.note(nextSteps, 'Next Steps');

        p.outro(
            color.green(`✓ Successfully rebranded ${this.changes.length} files!`)
        );
    }
}

// Run
const cli = new RebrandCLI();
cli.run().catch((error) => {
    console.error(color.red('Error:'), error.message);
    process.exit(1);
});
